stratum/lib/security/credentials_manager.cc

```diff
@@ -1,130 +1,161 @@
 // Copyright 2018 Google LLC
 // Copyright 2018-present Open Networking Foundation
+// Copyright 2023 Intel Corporation
 // SPDX-License-Identifier: Apache-2.0
 
 #include "stratum/lib/security/credentials_manager.h"
 
 #include <memory>
 #include <string>
 #include <utility>
+#include <fstream>
+#include <iostream>
 
 #include "absl/memory/memory.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
 #include "gflags/gflags.h"
-#include "grpcpp/security/server_credentials.h"
-#include "grpcpp/security/tls_credentials_options.h"
 #include "stratum/glue/logging.h"
-#include "stratum/glue/status/status.h"
 #include "stratum/lib/macros.h"
 #include "stratum/lib/utils.h"
 
-DEFINE_string(ca_cert, "", "CA certificate path");
-DEFINE_string(server_key, "", "gRPC Server pricate key path");
-DEFINE_string(server_cert, "", "gRPC Server certificate path");
+DEFINE_string(ca_cert_file, "", "Path to CA certificate file");
+DEFINE_string(server_key_file, "", "Path to gRPC server private key file");
+DEFINE_string(server_cert_file, "", "Path to gRPC server certificate file");
+DEFINE_string(client_key_file, "", "Path to gRPC client key file");
+DEFINE_string(client_cert_file, "", "Path to gRPC client certificate file");
 
 namespace stratum {
 
-CredentialsManager::~CredentialsManager() {}
+using ::grpc::experimental::FileWatcherCertificateProvider;
+using ::grpc::experimental::TlsChannelCredentialsOptions;
+using ::grpc::experimental::TlsServerCredentials;
+using ::grpc::experimental::TlsServerCredentialsOptions;
+
+constexpr unsigned int CredentialsManager::kFileRefreshIntervalSeconds;
+
 CredentialsManager::CredentialsManager() {}
 
+CredentialsManager::~CredentialsManager() {}
+
+::util::StatusOr<std::unique_ptr<CredentialsManager>>
+CredentialsManager::CreateInstance(bool secure_only /*=false*/) {
+  auto instance = absl::WrapUnique(new CredentialsManager());
+  RETURN_IF_ERROR(instance->Initialize(secure_only));
+  return std::move(instance);
+}
+
 std::shared_ptr<::grpc::ServerCredentials>
 CredentialsManager::GenerateExternalFacingServerCredentials() const {
   return server_credentials_;
 }
 
-::util::StatusOr<std::unique_ptr<CredentialsManager>>
-CredentialsManager::CreateInstance() {
-  auto instance_ = absl::WrapUnique(new CredentialsManager());
-  RETURN_IF_ERROR(instance_->Initialize());
-  return std::move(instance_);
+std::shared_ptr<::grpc::ChannelCredentials>
+CredentialsManager::GenerateExternalFacingClientCredentials() const {
+  return client_credentials_;
 }
 
-::util::Status CredentialsManager::Initialize() {
-  if (FLAGS_ca_cert.empty() && FLAGS_server_key.empty() &&
-      FLAGS_server_cert.empty()) {
-    LOG(WARNING) << "Using insecure server credentials";
-    server_credentials_ = ::grpc::InsecureServerCredentials();
+::util::Status CredentialsManager::Initialize(bool secure_only) {
```

Comment:
1. Unlike the original version of the `Initialize` method, this version can fail. It should return error status when it fails.
2. This version of `Initialize` is considerably more complex than the original version. It should be decomposed into `InitializeServer` and `InitializeClient` methods.
3. We may need to be able to select which of the (client, server) modes is required to succeed. If we're bringing up the server, does it matter that we can't initialize the client credentials, and vice versa? Can both modes be active at the same time? Perhaps the `CreateInstance` method should choose which initialization method(s) to invoke based on a parameter.
4. If we are unable to initialize the server credentials when we're running in server mode, the `infrap4d` process should exit with a suitable error message.

Reply:
Agree with everything you have mentioned here. However, I didnt want to get into re-writing this piece of the code and spending cycles in debugging and testing.

We can still use the `Initialize()` provided here and work around the problem described above. What I have done is that we call init() and only look for server_credentials_ (the client_credentials_ will be null). When used by the p4rt client or gnmi client, we only look if the client_credentials_ are valid.

We can look at improving this in the future for better control and better handling